R/filter_state_api_NEW.R

```diff
@@ -0,0 +1,442 @@
+#'
+filter_var <- function(
+    dataname,
+    varname,
+    choices = NULL,
+    selected = NULL,
+    keep_na = NULL,
+    keep_inf = NULL,
+    fixed = FALSE,
+    ...
+) {
+  checkmate::assert_string(dataname)
+  checkmate::assert_string(varname)
+  checkmate::assert_atomic(choices)
+  checkmate::assert_atomic(selected)
+  if (!is.null(choices) && !is.null(selected)) {
+    if (inherits(choices, c("integer", "numeric", "Date", "POSIXt"))) {
+      rc <- range(choices, na.rm = TRUE)
+      rs <- range(selected, na.rm = TRUE)
+      checkmate::assert_true(
+        rs[1L] >= rc[1L] && rs[2L] <= rc[2L],
+        .var.name = "range of \"selected\" is within range of \"choices\""
+      )
+    } else if (inherits(choices, c("logical", "character", "factor"))) {
+      checkmate::assert_subset(selected, choices)
+    } else {
+      stop("filter_var cannot handle \"choices\" of type: ", toString(class(choices)))
+    }
+  }
+  checkmate::assert_flag(keep_na, null.ok = TRUE)
+  checkmate::assert_flag(keep_inf, null.ok = TRUE)
+  checkmate::assert_flag(fixed)
+
+  ans <- list(
+    dataname = dataname,
+    varname = varname,
+    choices = choices,
+    selected = selected,
+    keep_na = keep_na,
+    keep_inf = keep_inf,
+    fixed = fixed
+  )
+  ans <- append(ans, list(...))
+
+  class(ans) <- c("teal_slice", class(ans))
+  ans
+}
+
+
+#' @export
+#' @rdname new_api
+#'
+filter_settings <- function(
+    ...,
+    exclude = list(),
+    count_type = c("none", "all", "hierarchical")
+) {
+  slices <- list(...)
+  checkmate::assert_list(slices, types = "teal_slice", any.missing = FALSE)
+  checkmate::assert_list(exclude, names = "named", types = "character")
+  count_type <- match.arg(count_type)
+
+  attr(slices, "exclude") <- exclude
+  attr(slices, "count_type") <- count_type
+  class(slices) <- c("teal_slices", class(slices))
+  slices
+}
+
+
+# check for teal_slice
+#' @export
+#' @rdname new_api
+#'
+is.teal_slice <- function(x) {
+  inherits(x, "teal_slice")
+}
+
+
+# convert list to teal_slice
+#' @export
+#' @rdname new_api
+#'
+as.teal_slice <- function(x) {
+  do.call(filter_var, x)
+}
+
+
+# concatenate method for teal_slice
+#' @export
+#' @rdname new_api
+#' @keywords internal
+#'
+c.teal_slice <- function(...) {
+  ans <- unlist(list(...), recursive = FALSE)
+  class(ans) <- "teal_slice"
+  ans
+}
+
+
+# print method for teal_slice
+#' @export
+#' @rdname new_api
+#' @keywords internal
+#'
+print.teal_slice <- function(x) {
+  name_width <- max(nchar(names(x)))
+  format_value <- function(v) {
+    if (is.null(v)) return("NULL")
+    if (inherits(v,  c("character", "factor"))) {
+      v <- dQuote(v, q = FALSE)
+    }
+    v <- paste(v, collapse = " ")
+    if (nchar(v) > 20L) {
+      v <- paste(substr(v, 1, 16), "...")
+    }
+    v
+  }
+
+  ind <- intersect(names(x), names(formals(filter_var)))
+  xx <- x[ind]
+  cat("teal_slice", "\n")
+  for (i in seq_along(xx)) {
+    element_name <- format(names(xx)[i], width = name_width)
```

Comment:
I wouldn't print if something was `NULL`
```suggestion
    if (is.null(xx[[i]])) next
    element_name <- format(names(xx)[i], width = name_width)
```

Reply:
Adding `show_all = FALSE` argument to `print` methods just in case.

Reply:
👍